pkg/core/region_tree.go

```diff
@@ -328,62 +328,106 @@ func (t *regionTree) getAdjacentItem(item *regionItem) (prev *regionItem, next *
 	return prev, next
 }
 
-// RandomRegion is used to get a random region within ranges.
-func (t *regionTree) RandomRegion(ranges []KeyRange) *RegionInfo {
-	if t.length() == 0 {
+func (t *regionTree) randomRegion(ranges []KeyRange) *RegionInfo {
+	regions := t.RandomRegions(1, ranges)
+	if len(regions) == 0 {
 		return nil
 	}
+	return regions[0]
+}
 
-	if len(ranges) == 0 {
-		ranges = []KeyRange{NewKeyRange("", "")}
+// RandomRegions get n random regions within the given ranges.
+func (t *regionTree) RandomRegions(n int, ranges []KeyRange) []*RegionInfo {
+	if t.length() == 0 || n < 1 {
+		return nil
 	}
-
-	for _, i := range rand.Perm(len(ranges)) {
-		var endIndex int
-		startKey, endKey := ranges[i].StartKey, ranges[i].EndKey
-		startRegion, startIndex := t.tree.GetWithIndex(&regionItem{RegionInfo: &RegionInfo{meta: &metapb.Region{StartKey: startKey}}})
-
-		if len(endKey) != 0 {
-			_, endIndex = t.tree.GetWithIndex(&regionItem{RegionInfo: &RegionInfo{meta: &metapb.Region{StartKey: endKey}}})
-		} else {
-			endIndex = t.tree.Len()
-		}
-
-		// Consider that the item in the tree may not be continuous,
-		// we need to check if the previous item contains the key.
-		if startIndex != 0 && startRegion == nil && t.tree.GetAt(startIndex-1).Contains(startKey) {
-			startIndex--
+	// Pre-allocate the variables to reduce the temporary memory allocations.
+	var (
+		startKey, endKey              []byte
+		startIndex, endIndex, randIdx int
+		startItem                     *regionItem
+		pivotItem                     = &regionItem{&RegionInfo{meta: &metapb.Region{}}}
+		region                        *RegionInfo
+		regions                       = make([]*RegionInfo, 0, n)
+		rangeNum, curLen              = len(ranges), len(regions)
+		// setStartEndIndices is a helper function to set `startIndex` and `endIndex`
+		// according to the `startKey` and `endKey`.
+		// TODO: maybe we could cache the `startIndex` and `endIndex` for each range.
+		setStartEndIndices = func() {
+			pivotItem.meta.StartKey = startKey
+			startItem, startIndex = t.tree.GetWithIndex(pivotItem)
+			if len(endKey) != 0 {
+				pivotItem.meta.StartKey = endKey
+				_, endIndex = t.tree.GetWithIndex(pivotItem)
+			} else {
+				endIndex = t.tree.Len()
+			}
+			// Consider that the item in the tree may not be continuous,
+			// we need to check if the previous item contains the key.
+			if startIndex != 0 && startItem == nil && t.tree.GetAt(startIndex-1).Contains(startKey) {
+				startIndex--
+			}
 		}
-
+	)
+	// If no ranges specified, select randomly from the whole tree.
+	// This is a fast path to reduce the unnecessary iterations.
+	if rangeNum == 0 {
+		startKey, endKey = []byte(""), []byte("")
+		setStartEndIndices()
 		if endIndex <= startIndex {
 			if len(endKey) > 0 && bytes.Compare(startKey, endKey) > 0 {
```

Comment:
```suggestion
			if len(endKey) > 0 && bytes.Compare(startKey, endKey) > 0 {
```
Is it possible to remove this `if condition` when the start and end keys are set to `”“`?

Reply:
Nice catch! Removed this check in this branch.